### PR TITLE
5014 editable tables

### PR DESCRIPTION
**Pull request by emitche**
Add products and fixed cost tables to loan questionnaire

### CODE REVIEW

File: app/views/admin/loans/content/_questionnaire_group.html.slim [hunk 1]
```diff
@@ -96,52 +96,17 @@
               = f.input_field "#{field.attribute_sym}__boolean".to_sym,
                 collection: %i(yes no), label_method: ->(i) { t("reply_#{i}") },
                 prompt: I18n.t("select_prompt"), data: {boolean: true}
+            
+            / Placeholder condition until breakeven data type is created
+            - if field.id == (304 || 90)
```
Comment (emitche) [inline app/views/admin/loans/content/_questionnaire_group.html.slim]: This is the wrong condition, but is temporary.

File: app/views/admin/loans/content/_table_question.html.slim [hunk 2]
```diff
@@ -0,0 +1,89 @@
+.table-responsive.editable-tables
+
+  / TODO: Convert input to expected input for loan question
+  input type="hidden" data-container="breakeven"
```
Comment (emitche) [inline app/views/admin/loans/content/_table_question.html.slim]: Fuzzy will adjust this to the correct field once the model is available.

File: app/views/admin/loans/content/_table_question.html.slim [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+          th.wide = t("loan.breakeven.description")
+          th = t("loan.breakeven.unit")
+          th.narrow = t("loan.breakeven.price")
+          th.narrow = t("loan.breakeven.cost")
+          th.narrow = t("loan.breakeven.quantity")
+          th.narrow: .actions: a data-action="add": i.fa.fa-plus-circle
+      tbody
+        / Hidden row is used to generate new rows
+        / Is at top instead of bottom to not affect striped rows
+        tr.ui-state-default.hidden data-group="product"
```
Comment (emitche) [inline app/views/admin/loans/content/_table_question.html.slim]: Ideally I would not be repeating myself and these would be separate partials. But then we might end up with lots of tiny partials once there are more tables. I like tiny partials, but know that isn't preferred in this project.

However, there are 3 groups of rows. One is hidden as described above. One is for actual data that is being loaded into the form. And one is an empty row to start out with whether or not the table has any existing data.

File: app/views/admin/loans/content/_table_question.html.slim [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            td: textarea data-input="unit" = product[:unit]
+            td: input data-input="price" value== product[:price] type="number"
+            td: input data-input="cost" value== product[:cost] type="number"
+            td: input data-input="quantity" value== product[:quantity] type="number"
+            td: .actions
+              a.delete data-action="delete": i.fa.fa-trash
+              a.move data-action="move": i.fa.fa-arrows-v
+
+        tr.ui-state-default data-group="product"
+          td: textarea data-input="name" placeholder="Name of product"
```
Comment (emitche) [inline app/views/admin/loans/content/_table_question.html.slim]: Placeholders and other are not yet translated. I am not sure if placeholders are desired and will check in with Roxana regarding visual review of feature.

File: app/assets/javascripts/backbone/views/editable_table_view.js.coffee [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+      masterInputValue = JSON.parse($masterInput.val())
+      masterInputValue["#{tableKey}"] = tableData
+      $masterInput.val(JSON.stringify(masterInputValue))
+
+  formatFixedCostsInput: ($row) ->
+    name = $row.find('[data-input="name"]').val()
+    amount = $row.find('[data-input="amount"]').val()
+
+    # Only format rows that have a name and amount
+    if Boolean(name) && Boolean(amount)
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/editable_table_view.js.coffee]: I haven't seen many folks doing this. Can you just do `if name && amount`? Empty string evaluates to false in Javascript.

File: app/assets/javascripts/backbone/views/editable_table_view.js.coffee [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    $table.append($new_row)
+
+  removeRow: (e) ->
+    e.preventDefault()
+    $row = $(e.currentTarget).closest('tr')
+    $row.remove()
+
+  saveTableData: (e) ->
+    $section = $(e.currentTarget)
+    $tables = $section.find('.editable-table')
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/editable_table_view.js.coffee]: I like the use of $ :)

File: app/assets/javascripts/backbone/views/editable_table_view.js.coffee [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    # Save the new data to each master input
+    $section.find('.editable-table').each (index) ->
+      $table = $(this)
+      tableKey = $table.data('table')
+      $rows = $table.find('tbody').find('tr')
+      tableData = []
+
+      for key,row of $rows
+        if !isNaN(key)
+          $row = $(row)
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/editable_table_view.js.coffee]: Please change to jquery each.

File: app/assets/javascripts/backbone/views/editable_table_view.js.coffee [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+          if rowResponse.rowData
+            tableData.push(rowResponse.rowData)
+
+      # Save generated table data to the master input, whose value is sent to server
+      $masterInput = $table.closest('.editable-tables').find('[data-container]')
+      masterInputValue = JSON.parse($masterInput.val())
+      masterInputValue["#{tableKey}"] = tableData
+      $masterInput.val(JSON.stringify(masterInputValue))
+
+  formatFixedCostsInput: ($row) ->
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/editable_table_view.js.coffee]: I think perhaps this should be called `prepareFixedCostsData`. `Format` suggests formatting stuff for display, to me. Same for the other method below.

File: app/views/admin/loans/content/_table_question.html.slim [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+          th = t("loan.breakeven.product_name")
+          th.wide = t("loan.breakeven.description")
+          th = t("loan.breakeven.unit")
+          th.narrow = t("loan.breakeven.price")
+          th.narrow = t("loan.breakeven.cost")
+          th.narrow = t("loan.breakeven.quantity")
+          th.narrow: .actions: a data-action="add": i.fa.fa-plus-circle
+      tbody
+        / Hidden row is used to generate new rows
+        / Is at top instead of bottom to not affect striped rows
```
Comment (hooverlunch) [inline app/views/admin/loans/content/_table_question.html.slim]: GREAT COMMENT.

File: app/views/admin/loans/content/_table_question.html.slim (hunk 3 above)
Comment (hooverlunch) [inline app/views/admin/loans/content/_table_question.html.slim]: Tiny partials are absolutely OK if they prevent code reuse. So please do refactor this.

File: app/assets/javascripts/backbone/views/editable_table_view.js.coffee [hunk 10]
```diff
@@ -0,0 +1,91 @@
+class MS.Views.EditableTableView extends Backbone.View
+
+  el: "table.editable-table"
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/editable_table_view.js.coffee]: Perhaps a comment here reminding that this view may control multiple elements, so not to save any element-specific state.

**Comment by hooverlunch** [comment]
Very nice work! I am impressed with the JSON building stuff. Just some small things and good to merge!

File: app/views/admin/loans/content/_table_question.html.slim (hunk 3 above)
Comment (emitche) [inline app/views/admin/loans/content/_table_question.html.slim]: Great! That makes me happy.

Comment (hooverlunch) [inline app/views/admin/loans/content/_table_question.html.slim]: Hooray for <sub>tiny partials</sub>

Comment (linesarefuzzy) [inline app/views/admin/loans/content/_table_question.html.slim]: FYI, I did this part. Sorry, Liz, it seems you were excited about it, but I thought I would be able to take care of all these changes. Alas.

Comment (hooverlunch) [inline app/views/admin/loans/content/_table_question.html.slim]: Something tells me she'll be fine with it :D

File: app/assets/javascripts/backbone/views/editable_table_view.js.coffee (hunk 5 above)
Comment (emitche) [inline app/assets/javascripts/backbone/views/editable_table_view.js.coffee]: This change has been made.

**Comment by emitche** [comment]
The items that needed to be addressed have been addressed in #143. Once that is merged, the open tag can be removed.

**Pull request merged**